mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala

```diff
@@ -27,10 +27,17 @@ import org.apache.spark.sql.DataFrame
 /**
  * Evaluator for multiclass classification.
  *
- * @param predictionAndLabels an RDD of (prediction, label) pairs.
+ * @param predAndLabelsWithOptWeight an RDD of (prediction, label, weight) or
+ *                         (prediction, label) pairs.
  */
 @Since("1.1.0")
-class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Double)]) {
+class MulticlassMetrics @Since("3.0.0") (predAndLabelsWithOptWeight: RDD[_]) {
```

Comment:
Oh, wait a sec, this changed the signature. I think you have to retain both. The `RDD[(Double, Double)]` constructor should stay, one way or the other, and add a new `RDD[(Double, Double, Double)]` constructor, with appropriate Since tags on each.

Below there's a `DataFrame` constructor and I'm not sure how to handle that. It should also handle the case where there's a weight col, but I'm not sure how to do that cleanly. There can be a second argument like `hasWeightCol` but that's starting to feel hacky.

Reply:
@srowen hmm, this was already suggested, please see this comment: https://github.com/apache/spark/pull/17086#discussion_r182303815
the build fails with an error due to Java type erasure, so this wouldn't work... you can't have two constructors with the same type erased signature... maybe I am misunderstanding something, and you meant something else?  Are you sure this changes the signature in a way that breaks others, it should still allow RDD with a tuple of 2 Double values.
The error I get is:
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala:35: double definition:
[error] constructor MulticlassMetrics: (predLabelsWeight: org.apache.spark.rdd.RDD[(Double, Double, Double)])org.apache.spark.mllib.evaluation.MulticlassMetrics at line 33 and
[error] constructor MulticlassMetrics: (predAndLabels: org.apache.spark.rdd.RDD[(Double, Double)])org.apache.spark.mllib.evaluation.MulticlassMetrics at line 35
[error] have same type after erasure: (predLabelsWeight: org.apache.spark.rdd.RDD)org.apache.spark.mllib.evaluation.MulticlassMetrics
[error] def this(predAndLabels: RDD[(Double, Double)]) =

Reply:
Darn, OK. Hm, so this doesn't actually cause a source or binary change? OK, that could be fine. I guess MiMa didn't complain. I guess you can now do weird things like pass `RDD[String]` here and it'll fail quickly. I'm a little uneasy about it but it's probably acceptable. Any other opinions?

I am not sure what to do about the DataFrame issue though. I suspect most people will want to call with a DataFrame now.

Reply:
"I am not sure what to do about the DataFrame issue though", ah, I think I see your concern.
But, isn't this dataframe constructor private anyway, so it can't be used by anyone outside mllib:

  private[mllib] def this(predictionAndLabels: DataFrame) =
    this(predictionAndLabels.rdd.map(r => (r.getDouble(0), r.getDouble(1))))

I only modified the RDD part because that is what is used by the ML evaluator and it is what users outside spark can access.  This is to add weight column for the evaluators.

However, even if we wanted to add weight column support for the private API, I'm unsure about how to add this.  Should I just check if there are 3 columns or two, and if there are 3 use the third one as the weight column?  I guess I am on the fence about this, I could change it but I don't think it is absolutely necessary, since it's not used anywhere outside spark MLLIB anyway.

Actually, this constructor is a bit weird, it looks like it was added as part of this PR:
https://github.com/apache/spark/pull/6011/files
It is only used here in the python API:
https://github.com/apache/spark/pull/6011/files#diff-443f766289f8090078531c3e1a1d6027R186
But I don't see why we couldn't just get the rdd there and remove the private constructor altogether (?)

Reply:
The python API takes an RDD, creates a DF, and then calls this private constructor with the DF, but I would think we could just pass the RDD directly